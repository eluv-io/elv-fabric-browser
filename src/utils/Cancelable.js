export const Cancelable = async (cancelable, f) => {
  if(cancelable) {
    return await cancelable.promise(() => f());
  } else {
    return await f();
  }
};
