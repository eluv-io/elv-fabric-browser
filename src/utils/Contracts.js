import BaseContentSpaceContract from "elv-client-js/src/contracts/BaseContentSpace";
import BaseContentContract from "elv-client-js/src/contracts/BaseContent";
import BaseContentTypeContract from "elv-client-js/src/contracts/BaseContentType";
import BaseLibraryContract from "elv-client-js/src/contracts/BaseLibrary";
import BaseAccessGroupContract from "elv-client-js/src/contracts/BaseAccessControlGroup";
import Fabric from "../clients/Fabric";

export const ContractTypes = {
  accessGroup: "accessGroup",
  contentSpace: "contentSpace",
  contentType: "contentType",
  customObject: "customObject",
  library: "library",
  object: "object",
  unknown: "unknown"
};

export const DetermineContractInterface = ({
  libraryId,
  objectId,
  isAccessGroup,
  isCustomContentObjectContract,
  contractAddressParam
}) => {
  const isContentSpaceLibrary = libraryId === Fabric.contentSpaceLibraryId;
  const isContentLibraryObject = objectId && (Fabric.utils.EqualHash(libraryId, objectId));

  if(isAccessGroup) {
    // Access Group
    return {
      type: ContractTypes.accessGroup,
      description: "Base Access Group Contract",
      abi: BaseAccessGroupContract.abi,
      contractAddress: contractAddressParam
    };
  }

  if(isContentSpaceLibrary && isContentLibraryObject) {
    // Content Space
    return {
      type: ContractTypes.contentSpace,
      description: "Base Content Space Contract",
      abi: BaseContentSpaceContract.abi,
      contractAddress: Fabric.utils.HashToAddress(libraryId)
    };
  }

  if(isContentLibraryObject) {
    // Library
    return {
      type: ContractTypes.library,
      description: "Base Content Library Contract",
      abi: BaseLibraryContract.abi,
      contractAddress: Fabric.utils.HashToAddress(libraryId)
    };
  }

  if(isContentSpaceLibrary && !isCustomContentObjectContract) {
    // Content Type
    return {
      type: ContractTypes.contentType,
      description: "Base Content Type Contract",
      abi: BaseContentTypeContract.abi,
      contractAddress: Fabric.utils.HashToAddress(objectId)
    };
  }

  if(objectId) {
    if(isCustomContentObjectContract) {
      // Custom content object or content type contract
      return {
        type: ContractTypes.customObject,
        description: "Custom Content Object Contract",
        abi: undefined,
        contractAddress: undefined
      };
    } else {
      // Content Object
      return {
        type: ContractTypes.object,
        description: "Base Content Object Contract",
        abi: BaseContentContract.abi,
        contractAddress: Fabric.utils.HashToAddress(objectId)
      };
    }
  }

  // Unknown
  return {
    type: ContractTypes.unknown,
    description: undefined,
    abi: undefined,
    contractAddress: contractAddressParam
  };
};
