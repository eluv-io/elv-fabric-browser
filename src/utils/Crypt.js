const Crypt = {
  strToArrayBuffer: (str) => {
    var buf = new ArrayBuffer(str.length * 2);
    var bufView = new Uint16Array(buf);
    for(var i = 0, strLen = str.length; i < strLen; i++) {
      bufView[i] = str.charCodeAt(i);
    }
    return buf;
  },

  arrayBufferToString: (buf) => {
    return String.fromCharCode.apply(null, new Uint16Array(buf));
  },

  generateKey: async (password) => {
    let utf8Password = new TextEncoder().encode(password);
    let passwordHash = await crypto.subtle.digest("SHA-256", utf8Password);

    let keygenAlgorithm = {
      name: "AES-GCM",
      length: 256
    };

    return crypto.subtle.importKey("raw", passwordHash, keygenAlgorithm, false, ["encrypt", "decrypt"]);
  },

  encrypt: async (plainText, password="bG9uZ2RlZmF1bHRwYXNzd29yZA") => {
    let iv = crypto.getRandomValues(new Uint8Array(12));
    let key = await Crypt.generateKey(password);

    let encryptionAlgorithm = {
      name: "AES-GCM",
      iv: iv,
      tagLength: 128
    };

    let encryptedText = await window.crypto.subtle.encrypt(encryptionAlgorithm, key, Crypt.strToArrayBuffer(plainText));

    return {
      iv: Crypt.arrayBufferToString(iv.buffer),
      encryptedText: Crypt.arrayBufferToString(encryptedText)
    };
  },

  decrypt: async (encryptedTextString, ivString, password="bG9uZ2RlZmF1bHRwYXNzd29yZA") => {
    let encryptedTextBuffer = Crypt.strToArrayBuffer(encryptedTextString);
    let iv = Crypt.strToArrayBuffer(ivString);
    let key = await Crypt.generateKey(password);

    let encryptionAlgorithm = {
      name: "AES-GCM",
      iv: iv,
      tagLength: 128
    };

    return Crypt.arrayBufferToString(
      await window.crypto.subtle.decrypt(encryptionAlgorithm, key, encryptedTextBuffer)
    );
  }
};

export default Crypt;
