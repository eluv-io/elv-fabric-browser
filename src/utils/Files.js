import UrlJoin from "url-join";

export const DownloadFromUrl = async (url, filename) => {
  let element = document.createElement("a");
  element.href = url;
  element.download = filename;

  element.style.display = "none";
  document.body.appendChild(element);

  element.click();

  document.body.removeChild(element);
  window.URL.revokeObjectURL(url);
};

// Convert a FileList to file info for UploadFiles
export const FileInfo = async (path, fileList, noData=false) => {
  return await Promise.all(
    Array.from(fileList).map(async file => {
      const data = noData ? undefined : await new Response(file).blob();
      const filePath = file.overrideName || file.webkitRelativePath || file.name;
      return {
        path: UrlJoin(path, filePath).replace(/^\/+/g, ""),
        type: "file",
        size: file.size,
        data
      };
    })
  );
};
