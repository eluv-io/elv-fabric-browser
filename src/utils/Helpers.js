// Traverse through a hashmap without throwing errors on undefined keys
// If any keys undefined, returns undefined
export const SafeTraverse = (object, ...keys) => {
  let result = object;

  for(let i = 0; i < keys.length; i++){
    result = result[keys[i]];

    if(result === undefined) { return undefined; }
  }

  return result;
};

export const Wait = async (ms) => {
  await new Promise(resolve => setTimeout(resolve, ms));
};