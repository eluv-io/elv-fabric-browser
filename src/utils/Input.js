// Properly handle input value updates
import React from "react";
import PropTypes from "prop-types";

export const InputValue = (event) => {
  if(event.target.type === "checkbox") {
    return event.target.checked;
  } else {
    return event.target.value;
  }
};

// Parse JSON for input values
// -- Allows whitespace and blank
// -- Rejects quoted strings which JSON.parse allows (e.g. JSON.parse('"string"')
export const ParseInputJson = (metadata) => {
  if(typeof metadata === "string") {
    metadata = metadata.trim();

    if(metadata === "") { return {}; }

    if(!metadata.startsWith("{") && !metadata.startsWith("[")) { throw Error("Invalid JSON"); }

    try {
      return JSON.parse(metadata);
    } catch(error) {
      throw Error("Invalid JSON");
    }
  }

  return metadata || {};
};

// Add to onKeyUp event to validate JSON input while typing
export const ValidateJsonInput = (value, target) => {
  try {
    const parsedValue = ParseInputJson(value);
    target.style.border = "";
    target.style.outline = "";
    target.title = "";
    return parsedValue;
  } catch(error) {
    target.style.border = "1px solid red";
    target.style.outline = "none";
    target.title = error;
  }
};

// Add to onBlur event to automatically format JSON input when
// focus is lost.
// UpdateValue is a function that updates the given value in component state
export const FormatJsonInput = (event, value, UpdateValue) => {
  const parsedValue = ValidateJsonInput(value, event.target);

  if(parsedValue) {
    UpdateValue(JSON.stringify(parsedValue, null, 2));
  }
};

// Automatically format and validate JSON
export const JsonTextArea = ({name, value, onChange, UpdateValue, className}) => {
  return (
    <textarea
      className={className}
      name={name}
      value={value}
      onChange={onChange}
      onKeyUp={(event) => {
        ValidateJsonInput(value, event.target);
      }}
      onBlur={(event) => FormatJsonInput(event, value, UpdateValue)}
    />
  );
};

JsonTextArea.propTypes = {
  name: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  UpdateValue: PropTypes.func.isRequired,
  className: PropTypes.string
};
