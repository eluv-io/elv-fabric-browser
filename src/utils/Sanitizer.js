class Sanitizer {
  static enforce_one(char, string){
    let first = string.indexOf(char) + 1;

    if(first === 0){ return string; }

    return string.substr(0, first) + string.slice(first).replace(char, "");
  }

  static enforce_first(char, string){
    if(string.length === 0){ return string; }

    return string.substr(0, 1) + string.slice(1).replace(char, "");
  }

  static s(value, type){
    switch(type.toLowerCase()){
      case("integer"):
        return value.toString().replace(/[^0-9]/gi,"");
      case("float"):
        return (
          this.enforce_first("-",
            this.enforce_one("-",
              this.enforce_one(".",
                value.toString().replace(/[^0-9\-\.]/gi,"")
              )
            )
          )
        );
      case("boolean"):
        return value === true || value === "true";
      default:
        return value;
    }
  }
}

export default Sanitizer;