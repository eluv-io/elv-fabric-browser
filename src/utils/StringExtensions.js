String.prototype.capitalize =
    function() {
      return this.replace(/_/g, " ").replace(/\w\S*/g, function (txt) {
        return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
      });
    };

String.prototype.toHash =
  function() {
    return this.split("").reduce((prevHash, currVal) =>
      (((prevHash << 5) - prevHash) + currVal.charCodeAt(0))|0, 0);
  };
