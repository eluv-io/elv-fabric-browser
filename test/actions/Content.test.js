import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import {
  CreateContentLibrary,
  GetFullContentObject,
  ListContentLibraries,
  ListContentObjects,

} from "../../src/actions/Content";

// Override Fabric library with mock version
import FabricMock from "../mocks/FabricRequester.mock";
import ElvWallet from "../mocks/ElvWallet.mock";
import {AwaitRequest} from "../utils";

const mockStore = configureStore([ thunk ]);
const store = mockStore();

// Listing and creating content libraries and objects
describe("Content Actions", () => {
  beforeEach(() => {
    store.clearActions();
    FabricMock.spy.clearSpyFunctions();
  });

  test("List content libraries", async () => {
    const requestId = store.dispatch(ListContentLibraries());

    expect(typeof requestId).toBe("number");

    expect(await AwaitRequest({ store, requestId })).toBe(true);

    expect(FabricMock.spy.ListContentLibraries).toHaveBeenCalledTimes(1);
    expect(FabricMock.spy.ListContentObjects).toHaveBeenCalledTimes(1);
    expect(store.getActions()).toMatchSnapshot();
  });

  test("List content objects", async () => {
    const requestId = store.dispatch(ListContentObjects({
      libraryId: "libraryId"
    }));

    expect(typeof requestId).toBe("number");

    expect(await AwaitRequest({ store, requestId })).toBe(true);

    expect(FabricMock.spy.ListContentObjects).toHaveBeenCalledTimes(1);
    expect(store.getActions()).toMatchSnapshot();
  });

  test("Get content object", async () => {
    const requestId = store.dispatch(GetFullContentObject({
      libraryId: "libraryId",
      contentHash: "contentHash"
    }));

    expect(typeof requestId).toBe("number");

    expect(await AwaitRequest({ store, requestId })).toBe(true);

    expect(FabricMock.spy.GetContentObjectMetadata).toHaveBeenCalledTimes(1);
    expect(store.getActions()).toMatchSnapshot();
  });

  test("Create content library", async () => {
    const wallet = new ElvWallet();

    const signer = wallet.AddAccount({
      accountName: "Alice",
      privateKey: "0x0000000000000000000000000000000000000000000000000000000000000000"
    });

    const requestId = store.dispatch(CreateContentLibrary({
      name: "Test content library",
      description: "Content library description",
      signer
    }));

    expect(typeof requestId).toBe("number");

    expect(await AwaitRequest({ store, requestId })).toBe(true);

    expect(FabricMock.spy.CreateContentLibrary).toHaveBeenCalledTimes(1);
    expect(store.getActions()).toMatchSnapshot();
  });
});
