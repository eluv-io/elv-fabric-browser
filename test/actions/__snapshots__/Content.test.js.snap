// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Actions Create content library 1`] = `
Array [
  Object {
    "action": "createContentLibrary",
    "domain": "content",
    "requestId": 4,
    "type": "REQUEST_SUBMITTED",
  },
  Object {
    "message": "Successfully created content library 'ilibPACFbXkPSawEKi7KnfwCfb'",
    "redirect": true,
    "type": "SET_NOTIFICATION_MESSAGE",
  },
  Object {
    "action": "createContentLibrary",
    "domain": "content",
    "requestId": 4,
    "type": "REQUEST_COMPLETED",
  },
]
`;

exports[`Content Actions Get content object 1`] = `
Array [
  Object {
    "action": "getContentObject",
    "domain": "content",
    "requestId": 3,
    "type": "REQUEST_SUBMITTED",
  },
  Object {
    "contentObject": ContentObject {
      "description": undefined,
      "hash": "hq__QmZPdxKM4wyvHG4En3LoAtZ9cQE9KL2MVNsEN17etCN2cA",
      "latestVersion": Object {
        "hash": "hq__QmZPdxKM4wyvHG4En3LoAtZ9cQE9KL2MVNsEN17etCN2cA",
        "id": "iq__8WCuZpLE6JjMKbLGa5KZaj",
        "meta": Object {
          "hash": "hq__QmZPdxKM4wyvHG4En3LoAtZ9cQE9KL2MVNsEN17etCN2cA",
          "id": "iq__6D1qLem852ggNcEYcAyvgT",
          "meta": Object {
            "genre": "Animation",
            "title": "Finding Dory",
            "year": 2016,
          },
          "type": "hq__QmRDQwpgAMkc3TDsnvwmv2VSgg3zqNNFcMWKSdxxMX4QxU",
        },
        "type": "hq__QmNyWraTrgY7rSdoRaejhedd7DBazwTAPGpSVYiZZQfAHw",
      },
      "libraryId": "libraryId",
      "metadata": Object {
        "hash": "hq__QmZPdxKM4wyvHG4En3LoAtZ9cQE9KL2MVNsEN17etCN2cA",
        "id": "iq__6D1qLem852ggNcEYcAyvgT",
        "meta": Object {
          "genre": "Animation",
          "title": "Finding Dory",
          "year": 2016,
        },
        "type": "hq__QmRDQwpgAMkc3TDsnvwmv2VSgg3zqNNFcMWKSdxxMX4QxU",
      },
      "name": "Content Object",
      "objectId": "iq__8WCuZpLE6JjMKbLGa5KZaj",
      "type": "Unknown",
      "versions": undefined,
    },
    "libraryId": "libraryId",
    "type": "COMPLETED_LIST_CONTENT_OBJECT",
  },
  Object {
    "action": "getContentObject",
    "domain": "content",
    "requestId": 3,
    "type": "REQUEST_COMPLETED",
  },
]
`;

exports[`Content Actions List content libraries 1`] = `
Array [
  Object {
    "action": "listContentLibraries",
    "domain": "content",
    "requestId": 1,
    "type": "REQUEST_SUBMITTED",
  },
  Object {
    "contentLibraries": Object {
      "ilibPACFbXkPSawEKi7KnfwCfb": ContentLibrary {
        "contentObjects": Array [
          ContentObject {
            "description": undefined,
            "hash": "hq__QmZPdxKM4wyvHG4En3LoAtZ9cQE9KL2MVNsEN17etCN2cA",
            "latestVersion": Object {
              "hash": "hq__QmZPdxKM4wyvHG4En3LoAtZ9cQE9KL2MVNsEN17etCN2cA",
              "id": "iq__6D1qLem852ggNcEYcAyvgT",
              "meta": Object {
                "genre": "Animation",
                "title": "Finding Dory",
                "year": 2016,
              },
              "type": "hq__QmRDQwpgAMkc3TDsnvwmv2VSgg3zqNNFcMWKSdxxMX4QxU",
            },
            "libraryId": "ilibPACFbXkPSawEKi7KnfwCfb",
            "metadata": Object {
              "genre": "Animation",
              "title": "Finding Dory",
              "year": 2016,
            },
            "name": "Content Object",
            "objectId": "iq__6D1qLem852ggNcEYcAyvgT",
            "type": "Unknown",
            "versions": Array [
              undefined,
            ],
          },
          ContentObject {
            "description": undefined,
            "hash": "hq__QmRDQwpgAMkc3TDsnvwmv2VSgg3zqNNFcMWKSdxxMX4QxU",
            "latestVersion": Object {
              "hash": "hq__QmRDQwpgAMkc3TDsnvwmv2VSgg3zqNNFcMWKSdxxMX4QxU",
              "id": "iq__L9LkebHK3QQ7q3rWxmWs55",
              "meta": Object {
                "bitcode": "video.bc",
              },
              "type": "",
            },
            "libraryId": "ilibPACFbXkPSawEKi7KnfwCfb",
            "metadata": Object {
              "bitcode": "video.bc",
            },
            "name": "Content Object",
            "objectId": "iq__L9LkebHK3QQ7q3rWxmWs55",
            "type": "Unknown",
            "versions": Array [
              undefined,
            ],
          },
        ],
        "description": "Hit the ground running this proposal is a win-win situation which will cause a stellar paradigm shift, and produce a multi-fold increase in deliverables but value-added. Pixel pushing i don't want to drain the whole swamp, i just want to shoot some alligators but I have zero cycles for this high-level it's a simple lift and shift job we need to have a Come to Jesus meeting with Phil about his attitude I just wanted to give you a heads-up. You better eat a reality sandwich before you walk back in that boardroom design thinking. Out of scope gain traction globalize idea shower critical mass. Streamline hit the ground running, or good optics. Wiggle room screw the pooch, for powerpoint Bunny. Shoot me an email fire up your browser, deliverables draw a line in the sand, yet nail jelly to the hothouse wall, and what's the status on the deliverables for eow? for upstream selling. Value-added horsehead offer. Not enough bandwidth show pony, so if you want to motivate these clowns, try less carrot and more stick, and red flag, or who's responsible for the ask for this request?. Organic growth wheelhouse not the long pole in my tent. We're ahead of the curve on that one we're ahead of the curve on that one. Quarterly sales are at an all-time low the last person we talked to said this would be ready Q1 not the long pole in my tent. Re-inventing the wheel. This vendor is incompetent . Meeting assassin viral engagement, but blue sky but cannibalize in an ideal world and it's about managing expectations. What's the status on the deliverables for eow? meeting assassin. Guerrilla marketing race without a finish line, or cross-pollination circle back ladder up / ladder back to the strategy digital literacy. Run it up the flagpole, ping the boss and circle back. Beef up Q1. Back to the drawing-board we need to touch base off-line before we fire the new ux experience so work flows , for are there any leftovers in the kitchen?. ",
        "libraryId": "ilibPACFbXkPSawEKi7KnfwCfb",
        "name": "Content Library",
      },
    },
    "type": "COMPLETED_LIST_ALL_CONTENT",
  },
  Object {
    "action": "listContentLibraries",
    "domain": "content",
    "requestId": 1,
    "type": "REQUEST_COMPLETED",
  },
]
`;

exports[`Content Actions List content objects 1`] = `
Array [
  Object {
    "action": "listContentObjects",
    "domain": "content",
    "requestId": 2,
    "type": "REQUEST_SUBMITTED",
  },
  Object {
    "contentLibrary": ContentLibrary {
      "contentObjects": Array [
        ContentObject {
          "description": undefined,
          "hash": "hq__QmZPdxKM4wyvHG4En3LoAtZ9cQE9KL2MVNsEN17etCN2cA",
          "latestVersion": Object {
            "hash": "hq__QmZPdxKM4wyvHG4En3LoAtZ9cQE9KL2MVNsEN17etCN2cA",
            "id": "iq__6D1qLem852ggNcEYcAyvgT",
            "meta": Object {
              "genre": "Animation",
              "title": "Finding Dory",
              "year": 2016,
            },
            "type": "hq__QmRDQwpgAMkc3TDsnvwmv2VSgg3zqNNFcMWKSdxxMX4QxU",
          },
          "libraryId": "libraryId",
          "metadata": Object {
            "genre": "Animation",
            "title": "Finding Dory",
            "year": 2016,
          },
          "name": "Content Object",
          "objectId": "iq__6D1qLem852ggNcEYcAyvgT",
          "type": "Unknown",
          "versions": Array [
            undefined,
          ],
        },
        ContentObject {
          "description": undefined,
          "hash": "hq__QmRDQwpgAMkc3TDsnvwmv2VSgg3zqNNFcMWKSdxxMX4QxU",
          "latestVersion": Object {
            "hash": "hq__QmRDQwpgAMkc3TDsnvwmv2VSgg3zqNNFcMWKSdxxMX4QxU",
            "id": "iq__L9LkebHK3QQ7q3rWxmWs55",
            "meta": Object {
              "bitcode": "video.bc",
            },
            "type": "",
          },
          "libraryId": "libraryId",
          "metadata": Object {
            "bitcode": "video.bc",
          },
          "name": "Content Object",
          "objectId": "iq__L9LkebHK3QQ7q3rWxmWs55",
          "type": "Unknown",
          "versions": Array [
            undefined,
          ],
        },
      ],
      "description": "Hit the ground running this proposal is a win-win situation which will cause a stellar paradigm shift, and produce a multi-fold increase in deliverables but value-added. Pixel pushing i don't want to drain the whole swamp, i just want to shoot some alligators but I have zero cycles for this high-level it's a simple lift and shift job we need to have a Come to Jesus meeting with Phil about his attitude I just wanted to give you a heads-up. You better eat a reality sandwich before you walk back in that boardroom design thinking. Out of scope gain traction globalize idea shower critical mass. Streamline hit the ground running, or good optics. Wiggle room screw the pooch, for powerpoint Bunny. Shoot me an email fire up your browser, deliverables draw a line in the sand, yet nail jelly to the hothouse wall, and what's the status on the deliverables for eow? for upstream selling. Value-added horsehead offer. Not enough bandwidth show pony, so if you want to motivate these clowns, try less carrot and more stick, and red flag, or who's responsible for the ask for this request?. Organic growth wheelhouse not the long pole in my tent. We're ahead of the curve on that one we're ahead of the curve on that one. Quarterly sales are at an all-time low the last person we talked to said this would be ready Q1 not the long pole in my tent. Re-inventing the wheel. This vendor is incompetent . Meeting assassin viral engagement, but blue sky but cannibalize in an ideal world and it's about managing expectations. What's the status on the deliverables for eow? meeting assassin. Guerrilla marketing race without a finish line, or cross-pollination circle back ladder up / ladder back to the strategy digital literacy. Run it up the flagpole, ping the boss and circle back. Beef up Q1. Back to the drawing-board we need to touch base off-line before we fire the new ux experience so work flows , for are there any leftovers in the kitchen?. ",
      "libraryId": "libraryId",
      "name": "Content Library",
    },
    "libraryId": "libraryId",
    "type": "COMPLETED_LIST_CONTENT_LIBRARY",
  },
  Object {
    "action": "listContentObjects",
    "domain": "content",
    "requestId": 2,
    "type": "REQUEST_COMPLETED",
  },
]
`;
